### PR TITLE
Improve overall performance

### PR DESCRIPTION
**Pull request by pscott**
This PR removes some redundant clones, uses more lazy `unwraps` / `or_insert`.

This basically is the result of running `cargo +nightly clippy -- -A clippy::all -D clippy::perf` on the whole substrate repo.

PS: a couple of `large_enum_variants` were omitted because these can sometimes be false negatives (we'd need to actually benchmark to be sure). Two `boxed_local` were omitted too because they came from a macro and I couldn't properly find where those came from.

### CODE REVIEW

**Comment by kianenigma** [comment]
would be interesting to use @NikVolf's bot to see if this makes a difference in the block times.

File: primitives/state-machine/src/changes_trie/changes_iterator.rs [hunk 1]
```diff
@@ -156,7 +156,7 @@ pub fn key_changes_proof_check_with_db<'a, H: Hasher, Number: BlockNumber>(
 	key: &[u8]
 ) -> Result<Vec<(Number, u32)>, String> where H::Out: Encode {
 	// we can't query any roots before root
-	let max = ::std::cmp::min(max.clone(), end.number.clone());
+	let max = ::std::cmp::min(max, end.number.clone());
```
Comment (bkchr) [inline primitives/state-machine/src/changes_trie/changes_iterator.rs:159]:
```suggestion
	let max = std::cmp::min(max, end.number.clone());
```

File: primitives/state-machine/src/changes_trie/changes_iterator.rs [hunk 2]
```diff
@@ -85,7 +85,7 @@ pub fn key_changes_proof<'a, H: Hasher, Number: BlockNumber>(
 	key: &[u8],
 ) -> Result<Vec<Vec<u8>>, String> where H::Out: Codec {
 	// we can't query any roots before root
-	let max = ::std::cmp::min(max.clone(), end.number.clone());
+	let max = ::std::cmp::min(max, end.number.clone());
```
Comment (bkchr) [inline primitives/state-machine/src/changes_trie/changes_iterator.rs:88]:
```suggestion
	let max = std::cmp::min(max, end.number.clone());
```

File: client/network/src/protocol/sync/extra_requests.rs [hunk 3]
```diff
@@ -141,7 +141,7 @@ impl<B: BlockT> ExtraRequests<B> {
 					request,
 				);
 			}
-			self.failed_requests.entry(request).or_insert(Vec::new()).push((who, Instant::now()));
+			self.failed_requests.entry(request).or_insert_with(|| Vec::new()).push((who, Instant::now()));
```
Comment (bkchr) [inline client/network/src/protocol/sync/extra_requests.rs:144]:
```suggestion
			self.failed_requests.entry(request).or_default().push((who, Instant::now()));
```

File: client/consensus/aura/src/lib.rs [hunk 4]
```diff
@@ -165,7 +165,7 @@ pub fn start_aura<B, C, SC, E, I, P, SO, CAW, Error>(
 	CAW: CanAuthorWith<B> + Send,
 {
 	let worker = AuraWorker {
-		client: client.clone(),
+		client: client,
```
Comment (bkchr) [inline client/consensus/aura/src/lib.rs:168]:
```suggestion
		client,
```

File: client/consensus/aura/src/lib.rs [hunk 5]
```diff
@@ -839,7 +839,7 @@ pub fn import_queue<B, I, C, P, S>(
 	initialize_authorities_cache(&*client)?;
 
 	let verifier = AuraVerifier {
-		client: client.clone(),
+		client: client,
```
Comment (bkchr) [inline client/consensus/aura/src/lib.rs:842]:
```suggestion
		client,
```

File: client/api/src/in_mem.rs [hunk 6]
```diff
@@ -155,7 +155,7 @@ impl<Block: BlockT> Blockchain<Block> {
 				aux: HashMap::new(),
 			}));
 		Blockchain {
-			storage: storage.clone(),
+			storage: storage,
```
Comment (bkchr) [inline client/api/src/in_mem.rs:158]:
```suggestion
			storage,
```

File: client/api/src/in_mem.rs [hunk 7]
```diff
@@ -124,7 +124,7 @@ impl<Block: BlockT + Clone> Clone for Blockchain<Block> {
 	fn clone(&self) -> Self {
 		let storage = Arc::new(RwLock::new(self.storage.read().clone()));
 		Blockchain {
-			storage: storage.clone(),
+			storage: storage,
```
Comment (bkchr) [inline client/api/src/in_mem.rs:127]:
```suggestion
			storage,
```

File: bin/node/cli/src/service.rs [hunk 8]
```diff
@@ -302,7 +302,7 @@ pub fn new_full_base(
 			inherent_data_providers: inherent_data_providers.clone(),
 			telemetry_on_connect: Some(telemetry_on_connect_sinks.on_connect_stream()),
 			voting_rule: grandpa::VotingRulesBuilder::default().build(),
-			prometheus_registry: prometheus_registry.clone(),
+			prometheus_registry: prometheus_registry,
```
Comment (bkchr) [inline bin/node/cli/src/service.rs:305]:
```suggestion
			prometheus_registry,
```

File: bin/node-template/node/src/service.rs [hunk 9]
```diff
@@ -182,11 +182,11 @@ pub fn new_full(config: Configuration) -> Result<TaskManager, ServiceError> {
...
 			config: grandpa_config,
 			link: grandpa_link,
-			network: network.clone(),
-			inherent_data_providers: inherent_data_providers.clone(),
+			network: network,
+			inherent_data_providers: inherent_data_providers,
 			telemetry_on_connect: Some(telemetry_on_connect_sinks.on_connect_stream()),
 			voting_rule: sc_finality_grandpa::VotingRulesBuilder::default().build(),
-			prometheus_registry: prometheus_registry.clone(),
+			prometheus_registry: prometheus_registry,
```
Comment (bkchr) [inline bin/node-template/node/src/service.rs:189]:
```suggestion
			prometheus_registry,
```

File: bin/node-template/node/src/service.rs [hunk 10]
```diff
@@ -182,11 +182,11 @@ pub fn new_full(config: Configuration) -> Result<TaskManager, ServiceError> {
 		let grandpa_config = sc_finality_grandpa::GrandpaParams {
 			config: grandpa_config,
 			link: grandpa_link,
-			network: network.clone(),
-			inherent_data_providers: inherent_data_providers.clone(),
+			network: network,
+			inherent_data_providers: inherent_data_providers,
```
Comment (bkchr) [inline bin/node-template/node/src/service.rs:186]:
```suggestion
			network,
			inherent_data_providers,
```

**Comment by pscott** [comment]
> > would be interesting to use @NikVolf's bot to see if this makes a difference in the block times.
> 
> I expect that you will not see any change at all. Most of the stuff doesn't touch any critical path or just switched from explicit `clone` to implicit `copy`.

I would agree! Definitely not expecting any major changes (if any changes at all) with this PR :) But I think it's still a `nicetohave` :)

File: primitives/state-machine/src/changes_trie/changes_iterator.rs [hunk 11]
```diff
@@ -46,7 +46,7 @@ pub fn key_changes<'a, H: Hasher, Number: BlockNumber>(
 	key: &'a [u8],
 ) -> Result<DrilldownIterator<'a, H, Number>, String> {
 	// we can't query any roots before root
-	let max = ::std::cmp::min(max.clone(), end.number.clone());
+	let max = ::std::cmp::min(max, end.number.clone());
```
Comment (kianenigma) [inline primitives/state-machine/src/changes_trie/changes_iterator.rs:49]:
```suggestion
	let max = std::cmp::min(max, end.number.clone());
```

**Pull request merged**